File: python-flask-api/app/controllers/teachers.py
from flask import Blueprint
from sqlalchemy import text
from app import db, config, jsonify, request
from ..helpers.http_errors import InternalServerError, BadRequest, ResourceNotFound
from ..helpers.utils import utils
from ..models.teachers import *


Teachers_blueprint = Blueprint('Teachers', __name__)


# list Teachers records
# @param fieldname: filter table records by a field
# @param fieldvalue:  filter value
# @request.args search: search records
# @request.args orderby: sort records by field name
# @request.args ordertype: sort type (asc|desc)
# @request.args page: current page number
# @request.args limit: maximum number of records to select
@Teachers_blueprint.route('/')
@Teachers_blueprint.route('/index/<fieldname>/<fieldvalue>')
def list(fieldname=None, fieldvalue=None):
    try:
        query = Teachers.query
        search = request.args.get('search')
        if search:
            query = query.filter(
                Teachers.fullname.like(f'%{search}%') | 
                Teachers.gender.like(f'%{search}%') | 
                Teachers.phone.like(f'%{search}%') | 
                Teachers.email.like(f'%{search}%') | 
                Teachers.id.like(f'%{search}%') 
            )
        
        # filter by dynamic field name
        if fieldname:
            field_filter = text(f'{fieldname} = :fieldvalue').params(fieldvalue=fieldvalue)
            query = query.filter(field_filter)
        
        orderby = request.args.get('orderby')
        ordertype = request.args.get('ordertype', 'desc')
        if orderby:
            query = query.order_by(text(f'{orderby} {ordertype}'))
        else:
            query = query.order_by(text(f'teachers.id {ordertype}'))
        
        # fields to select
        query = query.with_entities(
            Teachers.fullname,
            Teachers.dob,
            Teachers.gender,
            Teachers.phone,
            Teachers.email,
            Teachers.id
        )
        page = int(request.args.get('page', 1))
        limit = int(request.args.get('limit', config.MAX_RECORD_COUNT))
        offset = ((page-1) * limit)
        total_records = query.count()
        records = query.limit(limit).offset(offset).all()
        record_count = len(records)
        total_pages = round(total_records/limit)
        records = [(row._asdict()) for row in records]
        
        # response object
        response = dict(
            records=records, 
            totalRecords=total_records,
            recordCount=record_count,
            totalPages=total_pages
        )
        return jsonify(response)
    except Exception as ex:
        return InternalServerError(ex)


# Select table record by ID
@Teachers_blueprint.route('/view/<rec_id>')
def view(rec_id=None):
    try:
        query = Teachers.query
        query = query.filter(Teachers.id == rec_id)
        query = query.with_entities(
            Teachers.id,
            Teachers.fullname,
            Teachers.dob,
            Teachers.gender,
            Teachers.phone,
            Teachers.email
        )
        
        record = query.first()
        if not record: return ResourceNotFound()
         
        record = record._asdict()
        
        # return result as json
        return jsonify(record)
        
    except Exception as ex:
        return InternalServerError(ex)


# Save form data to the Teachers table
@Teachers_blueprint.route('/add', methods=['POST'])
def add():
    try:
        modeldata = request.body
        form = TeachersAddForm(modeldata)
        errors = [] # list of validation errors
        
        # validate teachers form data
        if not form.validate():
            errors.append(form.errors)
        
        if errors:
            return BadRequest(errors)
        
        record = Teachers()
        form.populate_obj(record)

        # check if phone record already exist in the database
        rec_value = str(modeldata['phone'])
        rec_exist = utils.is_unique(Teachers, "phone", rec_value)
        if rec_exist:
            return BadRequest(rec_value + " Already exist!")

        # check if email record already exist in the database
        rec_value = str(modeldata['email'])
        rec_exist = utils.is_unique(Teachers, "email", rec_value)
        if rec_exist:
            return BadRequest(rec_value + " Already exist!")
        
        # save teachers records
        db.session.add(record)
        db.session.commit()
        db.session.flush()
        rec_id = record.id
         
        record = record._asdict()
        return jsonify(record)
    except Exception as ex:
        return InternalServerError(ex)


# Select record by table primary key and update with form data
@Teachers_blueprint.route('/edit/<rec_id>', methods=['GET', 'POST'])
def edit(rec_id=None):
    try:
        query = Teachers.query
        query = query.filter(Teachers.id == rec_id)
        record = query.first()
        if not record: return ResourceNotFound()
        
        if request.method == 'POST':
            errors = []
            modeldata = request.body
            form = TeachersEditForm(modeldata, obj=record)
            
            if not form.validate():
                errors.append(form.errors)
            # check if phone already exist in the database
            rec_value = str(modeldata['phone'])
            rec_exist = utils.is_unique(Teachers, "phone", rec_value, "id", rec_id)
            if rec_exist:
                return BadRequest(rec_value  + " Already exist!")
            # check if email already exist in the database
            rec_value = str(modeldata['email'])
            rec_exist = utils.is_unique(Teachers, "email", rec_value, "id", rec_id)
            if rec_exist:
                return BadRequest(rec_value  + " Already exist!")
            
            if errors:
                return BadRequest(errors)
            
            # save Teachers record
            form.populate_obj(record)
            db.session.commit()
         
        record = record._asdict()
        return jsonify(record)
    except Exception as ex:
        return InternalServerError(ex)


# Delete record from the database
# Support multi delete by separating record id by comma.
@Teachers_blueprint.route('/delete/<rec_id>')
def delete(rec_id):
    query = Teachers.query
    arr_id = rec_id.split(',')
    try:
        query = query.filter(Teachers.id.in_(arr_id))
        query.delete(synchronize_session=False)
        db.session.commit()
        
        return jsonify(arr_id)
    except Exception as ex:
        return InternalServerError(ex)

File: nodejs-express-api/controllers/teachers.js
import { Router } from 'express';
import { body } from 'express-validator';
import validateFormData from '../helpers/validate_form.js';
import DB from '../models/db.js';


const router = Router();




/**
 * Route to list teachers records
 * @GET /teachers/index/{fieldname}/{fieldvalue}
 */
router.get(['/', '/index/:fieldname?/:fieldvalue?'], async (req, res) => {  
	try{
		const query = {};
		let queryFilters = [];
		let where = {};
		let replacements = {};
		let fieldName = req.params.fieldname;
		let fieldValue = req.params.fieldvalue;
		
		if (fieldName){
			queryFilters.push(DB.filterBy(fieldName, fieldValue));
		}
		let search = req.query.search;
		if(search){
			let searchFields = DB.Teachers.searchFields();
			where[DB.op.or] = searchFields;
			replacements.search = `%${search}%`;
		}
		
		if(queryFilters.length){
			where[DB.op.and] = queryFilters;
		}
		query.raw = true;
		query.where = where;
		query.replacements = replacements;
		query.order = DB.getOrderBy(req, 'id', 'desc');
		query.attributes = DB.Teachers.listFields();
		let page = parseInt(req.query.page) || 1;
		let limit = parseInt(req.query.limit) || 10;
		let result = await DB.Teachers.paginate(query, page, limit);
		return res.ok(result);
	}
	catch(err) {
		return res.serverError(err);
	}
});


/**
 * Route to view Teachers record
 * @GET /teachers/view/{recid}
 */
router.get('/view/:recid', async (req, res) => {
	try{
		const recid = req.params.recid || null;
		const query = {}
		const where = {}
		where['id'] = recid;
		query.raw = true;
		query.where = where;
		query.attributes = DB.Teachers.viewFields();
		let record = await DB.Teachers.findOne(query);
		if(!record){
			return res.notFound();
		}
		return res.ok(record);
	}
	catch(err){
		return res.serverError(err);
	}
});


/**
 * Route to insert Teachers record
 * @POST /teachers/add
 */
router.post('/add/', 
	[
		body('fullname').not().isEmpty(),
		body('dob').not().isEmpty(),
		body('gender').not().isEmpty(),
		body('phone').not().isEmpty(),
		body('email').not().isEmpty().isEmail(),
	], validateFormData
, async function (req, res) {
	try{
		let modeldata = req.getValidFormData();
		
		// check if phone already exist.
		let phoneCount = await DB.Teachers.count({ where:{ 'phone': modeldata.phone } });
		if(phoneCount > 0){
			return res.badRequest(`${modeldata.phone} already exist.`);
		}
		
		// check if email already exist.
		let emailCount = await DB.Teachers.count({ where:{ 'email': modeldata.email } });
		if(emailCount > 0){
			return res.badRequest(`${modeldata.email} already exist.`);
		}
		
		//save Teachers record
		let record = await DB.Teachers.create(modeldata);
		//await record.reload(); //reload the record from database
		const recid =  record['id'];
		const newValues = JSON.stringify(record); 
		req.writeToAuditLog({ recid, oldValues: null, newValues });
		
		return res.ok(record);
	} catch(err){
		return res.serverError(err);
	}
});


/**
 * Route to get  Teachers record for edit
 * @GET /teachers/edit/{recid}
 */
router.get('/edit/:recid', async (req, res) => {
	try{
		const recid = req.params.recid;
		const query = {};
		const where = {};
		where['id'] = recid;
		query.raw = true;
		query.where = where;
		query.attributes = DB.Teachers.editFields();
		let record = await DB.Teachers.findOne(query);
		if(!record){
			return res.notFound();
		}
		return res.ok(record);
	}
	catch(err){
		return res.serverError(err);
	}
});


/**
 * Route to update  Teachers record
 * @POST /teachers/edit/{recid}
 */
router.post('/edit/:recid', 
	[
		body('fullname').optional({nullable: true}).not().isEmpty(),
		body('dob').optional({nullable: true}).not().isEmpty(),
		body('gender').optional({nullable: true}).not().isEmpty(),
		body('phone').optional({nullable: true}).not().isEmpty(),
		body('email').optional({nullable: true}).not().isEmpty().isEmail(),
	], validateFormData
, async (req, res) => {
	try{
		const recid = req.params.recid;
		let modeldata = req.getValidFormData({ includeOptionals: true });
		
		// check if phone already exist.
		let phoneCount = await DB.Teachers.count({where:{'phone': modeldata.phone, 'id': {[DB.op.ne]: recid} }});
		if(phoneCount > 0){
			return res.badRequest(`${modeldata.phone} already exist.`);
		}
		
		// check if email already exist.
		let emailCount = await DB.Teachers.count({where:{'email': modeldata.email, 'id': {[DB.op.ne]: recid} }});
		if(emailCount > 0){
			return res.badRequest(`${modeldata.email} already exist.`);
		}
		const query = {};
		const where = {};
		where['id'] = recid;
		query.raw = true;
		query.where = where;
		query.attributes = DB.Teachers.editFields();
		let record = await DB.Teachers.findOne(query);
		if(!record){
			return res.notFound();
		}
		const oldValues = JSON.stringify(record); //for audit trail
		await DB.Teachers.update(modeldata, {where: where});
		record = await DB.Teachers.findOne(query);//for audit trail
		const newValues = JSON.stringify(record); 
		req.writeToAuditLog({ recid, oldValues, newValues });
		return res.ok(modeldata);
	}
	catch(err){
		return res.serverError(err);
	}
});


/**
 * Route to delete Teachers record by table primary key
 * Multi delete supported by recid separated by comma(,)
 * @GET /teachers/delete/{recid}
 */
router.get('/delete/:recid', async (req, res) => {
	try{
		const recid = (req.params.recid || '').split(',');
		const query = {};
		const where = {};
		where['id'] = recid;
		query.raw = true;
		query.where = where;
		let records = await DB.Teachers.findAll(query);
		records.forEach(async (record) => { 
			//perform action on each record before delete
			const oldValues = JSON.stringify(record); //for audit trail
			req.writeToAuditLog({ recid: record['id'], oldValues });
		});
		await DB.Teachers.destroy(query);
		return res.ok(recid);
	}
	catch(err){
		return res.serverError(err);
	}
});
export default router;
